qucs-filter/qucsfilter.1
.TH QucsFilter "1" "April 2005" "Debian/GNU Linux" "User Commands"
.SH NAME
QucsFilter \- A filter synthesis application.
.SH SYNOPSIS
.B qucsfilter
[\fIOPTION\fR] ...
.SH DESCRIPTION

\fBQucs\fR is an integrated circuit simulator which means you will be
able to setup a circuit with a graphical user interface (GUI) and
simulate the large-signal, small-signal and noise behaviour of the
circuit.  After that simulation has finished you will be able to
present the simulation results on a presentation page or window.

The software aims to support all kinds of circuit simulation types,
e.g. DC, AC, S-parameter, harmonic balance analysis, noise analysis,
etc.

\fBQucsFilter\fR is the filter synthesis tool used by Qucs.  By use of
an input dialog the user can create a filter which is then copied into
the system-wide clipboard.  In \fBQucs\fR the user opens an empty
schematic and presses CTRL-V (paste from clipboard). The filter
schematic is now inserted and can be simulated.

Available filter types are: Butterworth, Bessel, Chebyshev.

.SH AVAILABILITY
The latest version of Qucs can always be obtained from
\fBwww.sourceforge.net\fR or \fBwww.freshmeat.net\fR
.SH "REPORTING BUGS"
Known bugs are documented within the BUGS file.  Report bugs to
<qucs-bugs@lists.sourceforge.net>.
.SH COPYRIGHT
Copyright \(co 2005 Michael Margraf <michael.margraf@alumni.tu-berlin.de>
.PP
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH AUTHORS
Written by Toyoyuki Ishikawa <ishikawa@conceptkeeper.net> and Michael
Margraf <michael.margraf@alumni.tu-berlin.de>.


NEWS
--
-- NEWS
--
-- Copyright (C) 2003, 2004, 2005 Stefan Jahn <stefan@lkcc.org>
--
-- This is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2, or (at your option)
-- any later version.
--
-- This software is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this package; see the file COPYING.  If not, write to
-- the Free Software Foundation, Inc., 51 Franklin Street - Fifth Floor,
-- Boston, MA 02110-1301, USA.
--

This file lists user visible changes that were made between releases.  For
more verbose descriptions of these and internal changes see the ChangeLog
files.

Version 0.0.7
-------------

 * text painting rotations in subcircuit symbols
 * renamed Qucs, new name is "Quite Universal Circuit Simulator"
 * several small improvements and bug fixes
 * Latex-like mnemonics (e.g. \tau) in graphic texts and diagram labels
 * library manager including transistor, diode, substrate, etc. libraries

Version 0.0.6
-------------

 * finished SPICE component
 * transmission line calculator
 * filter synthesis application
 * new diagrams: 3-dimensional cartesian, locus curve
 * many small improvements
 * edit component properties directly on the schematic
 * translations into Swedish
 * new components: coplanar open, short, gap and step
 * nicer icons for simulation components

Version 0.0.5
-------------

 * internal help system now also in German available
 * translations into Hebrew, French, Japanese and Hungarian
 * some new components and diagrams
 * some new graph styles (stars, circles, ...)
 * manual axes scale for diagrams
 * many minor improvements and bug fixes
 * improved simulation component dialog
 * scroll bar for tabular diagrams
 * menu on right mouse button

Version 0.0.4
-------------

 * per-language internal help system implemented
 * translations into portuguese and spanish
 * some new paintings and components
 * subcircuits with symbol painted by the user
 * improved the zooming (looks nicer now)
 * progress bar during simulation
 * some small improvements and bug fixes

Version 0.0.3
-------------

 * avoided working directory changes during program execution
 * translations into Polish and Romanian
 * some new components (e.g. transistors)
 * some new operations (e.g. aligning)
 * undo/redo function
 * improved subcircuit support
 * resize operation for paintings
 * logarithmical scale for rectangle diagrams
 * change style of graphs (dashed, dotted, ...)
 * qucs-edit and qucs-help as external programs
 * application settings dialog
 * many small improvements and bug fixes

Version 0.0.2
-------------

 * translations into german and italian
 * modified build system, no more kdevelop
 * a (very modest) help system
 * deleting and rename of documents
 * some new components and paintings
 * runs more stable
 * mouse cursor symbols show active action
 * document area more flexible for resize and zoom
 * equation support
 * strongly improved handling with wire labels
 * diagrams more beautiful
 * diagrams can show independent and more-dimensional graphs
 * markers can be set on graphs

Version 0.0.1
-------------

 * first available version


qucs-filter/qucsfilter.cpp
/***************************************************************************
                               qucsfilter.cpp
                             ------------------
    begin                : Wed Mar 02 2005
    copyright            : (C) 2005 by Michael Margraf
    email                : michael.margraf@alumni.tu-berlin.de
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include "qucsfilter.h"
#include "helpdialog.h"
#include "lc_filter.h"

#include <qmenubar.h>
#include <qpopupmenu.h>
#include <qmessagebox.h>
#include <qlayout.h>
#include <qlabel.h>
#include <qpushbutton.h>
#include <qlineedit.h>
#include <qcombobox.h>
#include <qvalidator.h>
#include <qtimer.h>
#include <qclipboard.h>
#include <qapplication.h>

#include <math.h>




QucsFilter::QucsFilter()
{
  // set application icon
  setIcon (QPixmap(QucsSettings.BitmapDir + "big.qucs.xpm"));
  setCaption("QucsFilter " PACKAGE_VERSION);


  // --------  create menubar  -------------------
  QPopupMenu *fileMenu = new QPopupMenu();
  fileMenu->insertItem(tr("E&xit"), this, SLOT(slotQuit()), CTRL+Key_Q);

  QPopupMenu *helpMenu = new QPopupMenu();
  helpMenu->insertItem(tr("Help..."), this, SLOT(slotHelpIntro()), Key_F1);
  helpMenu->insertSeparator();
  helpMenu->insertItem(
                tr("&About QucsFilter..."), this, SLOT(slotHelpAbout()), 0);
  helpMenu->insertItem(tr("About Qt..."), this, SLOT(slotHelpAboutQt()), 0);

  QMenuBar *bar = new QMenuBar(this);
  bar->insertItem(tr("&File"), fileMenu);
  bar->insertItem(tr("&Help"), helpMenu);


  // -------  create main windows widgets --------
  gbox = new QGridLayout(this, 10,3,5,5);

  QWidget *Space = new QWidget(this);   // reserve space for menubar
  Space->setFixedSize(5, bar->height());
  gbox->addWidget(Space, 0,0);

  QLabel *Label1 = new QLabel(tr("Filter type:"), this);
  gbox->addWidget(Label1, 1,0);
  ComboType = new QComboBox(this);
  ComboType->insertItem("Bessel");
  ComboType->insertItem("Butterworth");
  ComboType->insertItem("Chebyshev");
  gbox->addWidget(ComboType, 1,1);
  connect(ComboType, SIGNAL(activated(int)), SLOT(slotTypeChanged(int)));

  QLabel *Label2 = new QLabel(tr("Filter class:"), this);
  gbox->addWidget(Label2, 2,0);
  ComboClass = new QComboBox(this);
  ComboClass->insertItem(tr("Low pass"));
  ComboClass->insertItem(tr("High pass"));
  ComboClass->insertItem(tr("Band pass"));
  ComboClass->insertItem(tr("Band stop"));
  gbox->addWidget(ComboClass, 2,1);
  connect(ComboClass, SIGNAL(activated(int)), SLOT(slotClassChanged(int)));

  IntVal = new QIntValidator(1, 200, this);
  DoubleVal = new QDoubleValidator(this);

  QLabel *Label3 = new QLabel(tr("Order:"), this);
  gbox->addWidget(Label3, 3,0);
  EditOrder = new QLineEdit("3", this);
  EditOrder->setValidator(IntVal);
  gbox->addWidget(EditOrder, 3,1);

  LabelStart = new QLabel(tr("Corner frequency:"), this);
  gbox->addWidget(LabelStart, 4,0);
  EditCorner = new QLineEdit("1", this);
  EditCorner->setValidator(DoubleVal);
  gbox->addWidget(EditCorner, 4,1);
  ComboCorner = new QComboBox(this);
  ComboCorner->insertItem("Hz");
  ComboCorner->insertItem("kHz");
  ComboCorner->insertItem("MHz");
  ComboCorner->insertItem("GHz");
  ComboCorner->setCurrentItem(3);
  gbox->addWidget(ComboCorner, 4,2);

  LabelStop = new QLabel(tr("Stop frequency:"), this);
  gbox->addWidget(LabelStop, 5,0);
  EditStop = new QLineEdit("2", this);
  EditStop->setValidator(DoubleVal);
  gbox->addWidget(EditStop, 5,1);
  ComboStop = new QComboBox(this);
  ComboStop->insertItem("Hz");
  ComboStop->insertItem("kHz");
  ComboStop->insertItem("MHz");
  ComboStop->insertItem("GHz");
  ComboStop->setCurrentItem(3);
  gbox->addWidget(ComboStop, 5,2);

  LabelRipple = new QLabel(tr("Pass band ripple:"), this);
  gbox->addWidget(LabelRipple, 6,0);
  EditRipple = new QLineEdit("1", this);
  EditRipple->setValidator(DoubleVal);
  gbox->addWidget(EditRipple, 6,1);
  Label_dB = new QLabel("dB", this);
  gbox->addWidget(Label_dB, 6,2);

  QLabel *Label9 = new QLabel(tr("Impedance:"), this);
  gbox->addWidget(Label9, 7,0);
  EditImpedance = new QLineEdit("50", this);
  EditImpedance->setValidator(DoubleVal);
  gbox->addWidget(EditImpedance, 7,1);
  QLabel *Label10 = new QLabel("Ohm", this);
  gbox->addWidget(Label10, 7,2);


  QPushButton *ButtonGo =
               new QPushButton(tr("Calculate and put into Clipboard"), this);
  connect(ButtonGo, SIGNAL(clicked()), SLOT(slotCalculate()));
  gbox->addMultiCellWidget(ButtonGo, 8,8,0,2);

  LabelResult = new QLabel(this);
  ResultState = 100;
  slotShowResult();
  LabelResult->setAlignment(Qt::AlignHCenter);
  gbox->addMultiCellWidget(LabelResult, 9,9,0,2);


  // -------  finally set initial state  --------
  slotTypeChanged(0);
  slotClassChanged(0);
}

QucsFilter::~QucsFilter()
{
  delete gbox;
  delete IntVal;
  delete DoubleVal;
}

// ************************************************************
void QucsFilter::slotQuit()
{
  int tmp;
  tmp = x();		// call size and position function in order to ...
  tmp = y();		// ... set them correctly before closing the ...
  tmp = width();	// dialog !!!  Otherwise the frame of the window ...
  tmp = height();	// ... will not be recognized (a X11 problem).

  accept();
}

// ************************************************************
void QucsFilter::slotHelpAbout()
{
  QMessageBox::about(this, tr("About..."),
    tr("QucsFilter Version ")+PACKAGE_VERSION+
    tr("\nFilter synthesis program\n")+
    tr("Copyright (C) 2005 by")+
    "\n     Toyoyuki Ishikawa, Vincent Habchi, Michael Margraf\n"
    "\nThis is free software; see the source for copying conditions."
    "\nThere is NO warranty; not even for MERCHANTABILITY or "
    "\nFITNESS FOR A PARTICULAR PURPOSE.\n\n");
}

// ************************************************************
void QucsFilter::slotHelpAboutQt()
{
  QMessageBox::aboutQt(this, tr("About Qt"));
}

// ************************************************************
void QucsFilter::slotHelpIntro()
{
  HelpDialog *d = new HelpDialog(this);
  d->show();
}

// ************************************************************
void QucsFilter::setError(const QString& Message)
{
  LabelResult->setText(tr("Result:") + "<font color=\"#FF0000\"><b>  " +
                       tr("Error") + "</b></font>");
  QMessageBox::critical(this, tr("Error"), Message);
}

// ************************************************************
void QucsFilter::slotCalculate()
{
  // get numerical values from input widgets
  double CornerFreq = EditCorner->text().toDouble();
  double StopFreq   = EditStop->text().toDouble();

  CornerFreq *= pow(10, double(3*ComboCorner->currentItem())); // add exponent
  StopFreq   *= pow(10, double(3*ComboStop->currentItem()));

  tFilter Filter;
  Filter.Type = ComboType->currentItem();
  Filter.Class = ComboClass->currentItem();
  Filter.Order = EditOrder->text().toInt();
  Filter.Ripple = EditRipple->text().toDouble();
  Filter.Impedance = EditImpedance->text().toDouble();
  Filter.Frequency = CornerFreq;
  Filter.Frequency2 = StopFreq;

  if(EditStop->isEnabled())
    if(Filter.Frequency >= Filter.Frequency2) {
      setError(tr("Stop frequency must be greater than start frequency."));
      return;
    }

  if(Filter.Order < 2) {
    setError(tr("Filter order must not be less than two."));
    return;
  }

  QString *s = LC_Filter::createSchematic(&Filter);
  if(!s) return;


  // put resulting filter schematic into clipboard
  QClipboard *cb = QApplication::clipboard();
  cb->setText(*s);
  delete s;

  // show result for some time
  ResultState = 0;
  LabelResult->setText(tr("Result:") + "<font color=\"#008000\"><b>  " +
                       tr("Successful") + "</b></font>");
  QTimer::singleShot(500, this, SLOT(slotShowResult()));
}

// ************************************************************
void QucsFilter::slotShowResult()
{
  if(ResultState > 5) {
    LabelResult->setText(tr("Result: --"));
    return;
  }


  int c;
  ResultState++;
  if(ResultState & 1)  c = 0xFF;
  else c = 0x80;
  QString s = QString("<font color=\"#00%100\"><b>  ").arg(c, 2, 16);
  LabelResult->setText(tr("Result:") + s + tr("Successful") + "</b></font>");

  c = 500;
  if(ResultState > 5)  c = 3000;
  QTimer::singleShot(c, this, SLOT(slotShowResult()));
}

// ************************************************************
void QucsFilter::slotTypeChanged(int index)
{
  switch(index) {
    case TYPE_BESSEL :
    case TYPE_BUTTERWORTH :
	LabelRipple->setEnabled(false);
	EditRipple->setEnabled(false);
	Label_dB->setEnabled(false);
	break;
    case TYPE_CHEBYSHEV :
	LabelRipple->setEnabled(true);
	EditRipple->setEnabled(true);
	Label_dB->setEnabled(true);
	break;
  }
}

// ************************************************************
void QucsFilter::slotClassChanged(int index)
{
  switch(index) {
    case CLASS_LOWPASS :
    case CLASS_HIGHPASS :
	LabelStop->setEnabled(false);
	EditStop->setEnabled(false);
	ComboStop->setEnabled(false);
	LabelStart->setText(tr("Corner frequency:"));
	break;
    case CLASS_BANDPASS :
    case CLASS_BANDSTOP :
	LabelStop->setEnabled(true);
	EditStop->setEnabled(true);
	ComboStop->setEnabled(true);
	LabelStart->setText(tr("Start frequency:"));
	break;
  }
}
